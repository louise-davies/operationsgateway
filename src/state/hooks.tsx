import { useDispatch, useSelector } from 'react-redux';
import type { TypedUseSelectorHook } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { createSelector } from '@reduxjs/toolkit';

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

const everythingButConfigSelector = ({ config, ...state }: RootState) => state;
export const sessionSelector = createSelector(
  everythingButConfigSelector,
  (state) => state,
  {
    devModeChecks: {
      identityFunctionCheck: 'never',
      inputStabilityCheck: 'never',
    },
  }
);
