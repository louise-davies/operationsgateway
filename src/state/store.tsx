import { configureStore } from '@reduxjs/toolkit';
import OperationsGatewayMiddleware, {
  listenToMessages,
} from './middleware/operationsgateway.middleware';
import configReducer from './slices/configSlice';

export const store = configureStore({
  reducer: {
    config: configReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(OperationsGatewayMiddleware),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

listenToMessages(store.dispatch);
