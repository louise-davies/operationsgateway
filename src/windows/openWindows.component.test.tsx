import React from 'react';
import OpenWindows from './openWindows.component';
import {
  getInitialState,
  renderComponentWithStore,
  testPlotConfigs,
} from '../setupTests';
import { PreloadedState } from '@reduxjs/toolkit';
import { RootState } from '../state/store';

// need to mock to avoid errors
jest.mock('../plotting/plotWindow.component', () => (props) => (
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  <mock-plotWindow data-testid="mock-plotWindow">
    {Object.entries(props).map(
      ([propName, propValue]) =>
        `${propName}=${JSON.stringify(propValue, null, 2)}\n`
    )}
    {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
    {/* @ts-ignore */}
  </mock-plotWindow>
));

describe('Open Windows component', () => {
  let state: PreloadedState<RootState>;

  const createView = (initialState = state) => {
    return renderComponentWithStore(<OpenWindows />, {
      preloadedState: initialState,
    });
  };

  beforeEach(() => {
    state = getInitialState();
  });

  it('renders windows for each open plot', async () => {
    state.plots = {
      [testPlotConfigs[0].title]: testPlotConfigs[0],
      [testPlotConfigs[1].title]: testPlotConfigs[1],
      [testPlotConfigs[2].title]: testPlotConfigs[2],
    };
    const view = createView();

    // We expect Plot 0 and Plot 2 to be in the screenshot. Plot 1 has open: false
    expect(view.asFragment()).toMatchSnapshot();
  });
});
