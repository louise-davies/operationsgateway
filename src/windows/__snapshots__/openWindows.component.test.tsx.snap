// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Open Windows component renders windows for each open plot 1`] = `
<DocumentFragment>
  <mock-plotwindow
    data-testid="mock-plotWindow"
  >
    plotConfig={
  "open": true,
  "title": "Plot 0",
  "plotType": "scatter",
  "XAxis": "timestamp",
  "XAxisScale": "time",
  "selectedPlotChannels": [],
  "leftYAxisScale": "linear",
  "rightYAxisScale": "logarithmic",
  "gridVisible": true,
  "axesLabelsVisible": false,
  "selectedColours": [],
  "remainingColours": [
    "#008000",
    "#0000ff",
    "#ff00ff",
    "#00ffff",
    "#008080",
    "#800000",
    "#00ff00",
    "#000080",
    "#7f8000",
    "#80007f"
  ],
  "outerWidth": 600,
  "outerHeight": 400,
  "screenX": 200,
  "screenY": 200
}
onClose=undefined

  </mock-plotwindow>
  <mock-plotwindow
    data-testid="mock-plotWindow"
  >
    plotConfig={
  "open": true,
  "title": "Plot 2",
  "plotType": "scatter",
  "XAxisScale": "logarithmic",
  "selectedPlotChannels": [],
  "leftYAxisScale": "linear",
  "rightYAxisScale": "logarithmic",
  "gridVisible": true,
  "axesLabelsVisible": false,
  "selectedColours": [],
  "remainingColours": [
    "#008000",
    "#0000ff",
    "#ff00ff",
    "#00ffff",
    "#008080",
    "#800000",
    "#00ff00",
    "#000080",
    "#7f8000",
    "#80007f"
  ],
  "outerWidth": 600,
  "outerHeight": 400,
  "screenX": 200,
  "screenY": 200
}
onClose=undefined

  </mock-plotwindow>
</DocumentFragment>
`;
